mungegithub/mungers/approvers/owners.go

```diff
@@ -0,0 +1,200 @@
+	return Owners{filenames: filenames, repo: r, seed: s}
+}
+
+
+// removeSubdirs takes a list of directories as an input and returns a set of directories with all
+// subdirectories removed.  E.g. [/a,/a/b/c,/d/e,/d/e/f] -> [/a, /d/e]
+func removeSubdirs(dirList []string) sets.String {
+	toDel := sets.String{}
+	for i := 0; i < len(dirList)-1; i++ {
+		for j := i + 1; j < len(dirList); j++ {
+			// ex /a/b has prefix /a so if remove /a/b since its already covered
+			if strings.HasPrefix(dirList[i], dirList[j]) {
+				toDel.Insert(dirList[i])
+			} else if strings.HasPrefix(dirList[j], dirList[i]) {
+				toDel.Insert(dirList[j])
+			}
+		}
+	}
+	finalSet := sets.NewString(dirList...)
+	finalSet.Delete(toDel.List()...)
+	return finalSet
+}
+
+func (o Owners) GetApprovers() map[string]sets.String {
+	ownersToApprovers := map[string]sets.String{}
+
+	for _, fn := range o.ListOwnersFiles() {
+		ownersToApprovers[fn] = o.repo.LeafApprovers(fn)
+	}
+
+	return ownersToApprovers
+}
+
+func (o Owners) GetReverseMap() map[string]sets.String {
+	approverOwnersfiles := map[string]sets.String{}
+	for ownersFile, approvers := range o.GetApprovers() {
+		for approver := range approvers {
+			if _, ok := approverOwnersfiles[approver]; ok {
+				approverOwnersfiles[approver].Insert(ownersFile)
+			} else {
+				approverOwnersfiles[approver] = sets.NewString(ownersFile)
+			}
+		}
+	}
+	return approverOwnersfiles
+}
+
+func (o Owners) GetShuffledApprovers() (people []string, files []sets.String) {
+	approverOwnersfiles := o.GetReverseMap()
+	sliceOfKeys := make([]string, 0, len(approverOwnersfiles))
+	for approver := range approverOwnersfiles {
+		sliceOfKeys = append(sliceOfKeys, approver)
+	}
+	sort.Strings(sliceOfKeys)
+	order := rand.Perm(len(approverOwnersfiles))
+	people = make([]string, 0, len(approverOwnersfiles))
+	for i := range sliceOfKeys {
+		people = append(people, sliceOfKeys[order[i]])
+	}
+
+	files = []sets.String{}
+	for _, person := range people {
+		files = append(files, approverOwnersfiles[person])
+	}
+	return people, files
+}
+
+
+func (o Owners) GetSuggestedApprovers() []string {
+
+	randomizedApprovers, filesCanApprove := o.GetShuffledApprovers()
+
+	copyOfFiles := sets.NewString()
+	for _, fn := range o.filenames {
+		copyOfFiles.Insert(fn)
+	}
+
+	approverGroup := sets.NewString()
+	var bestPerson string
+	var bestPersonFiles sets.String
+	for copyOfFiles.Len() > 0 {
+		maxCovered := 0
+		for i, approver := range randomizedApprovers {
+			filesCanApprove := filesCanApprove[i]
+			if filesCanApprove.Intersection(copyOfFiles).Len() > maxCovered {
+				maxCovered = len(filesCanApprove)
+				bestPerson = approver
+				bestPersonFiles = filesCanApprove
+			}
+		}
+
+		approverGroup.Insert(bestPerson)
+		toDelete := sets.NewString()
+		// remove all files in the directories that our approver approved AND
+		// in the subdirectories that s/he approved.  HasPrefix finds subdirs
+		for fn := range copyOfFiles {
```

Comment:
I don't exactly get that.

At this point you have the suggested person, don't you just want to remove from `copyOfFiles` (which is a terrible name) the list of files that person can approve? `copyOfFiles = copyOfFiles.Difference(filesCanApprove[bestPerson])` and remove a bunch of temporary?